src/backend/distributed/planner/recursive_planning.c

```diff
@@ -1291,13 +1300,21 @@ ShouldRecursivelyPlanSubquery(Query *subquery, RecursivePlanningContext *context
 		 * subquery, we cannot push push it down and therefore we should try to
 		 * recursively plan it.
 		 */
-		if (!context->allDistributionKeysInQueryAreEqual &&
+		if (!context->allDistKeysInQueryAreEqualAndColocated &&
 			!AllDistributionKeysInSubqueryAreEqual(subquery,
 												   context->plannerRestrictionContext))
 		{
 			return true;
 		}
 
+		if (!context->allDistKeysInQueryAreEqualAndColocated &&
```

Comment:
Again, I cannot follow this here.

We are already capable of doing non-colocated subquery joins without this on non-colocated tables:

```SQL
explain select * from (SELECT * FROM dist_2) foo JOIN distributed_table as bar USING (a) LIMIT 1;
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                            QUERY PLAN                                                             │
├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Limit  (cost=0.00..0.00 rows=1 width=8)                                                                                           │
│   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                                                         │
│         ->  Distributed Subplan 12_1                                                                                              │
│               ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=4)                                             │
│                     Task Count: 32                                                                                                │
│                     Tasks Shown: One of 32                                                                                        │
│                     ->  Task                                                                                                      │
│                           Node: host=localhost port=9701 dbname=onderkalaci                                                       │
│                           ->  Seq Scan on dist_2_102047 dist_2  (cost=0.00..35.50 rows=2550 width=4)                              │
│         Task Count: 32                                                                                                            │
│         Tasks Shown: One of 32                                                                                                    │
│         ->  Task                                                                                                                  │
│               Node: host=localhost port=9701 dbname=onderkalaci                                                                   │
│               ->  Limit  (cost=0.00..3.01 rows=1 width=8)                                                                         │
│                     ->  Nested Loop  (cost=0.00..33948.25 rows=11300 width=8)                                                     │
│                           Join Filter: (intermediate_result.a = bar.a)                                                            │
│                           ->  Function Scan on read_intermediate_result intermediate_result  (cost=0.00..10.00 rows=1000 width=4) │
│                           ->  Materialize  (cost=0.00..43.90 rows=2260 width=8)                                                   │
│                                 ->  Seq Scan on distributed_table_102014 bar  (cost=0.00..32.60 rows=2260 width=8)                │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(19 rows)
```

where
```SQL
 select colocationid from pg_dist_partition WHERE logicalrelid IN ('distributed_table'::regclass, 'dist_2'::regclass);
┌──────────────┐
│ colocationid │
├──────────────┤
│            5 │
│            6 │
└──────────────┘
(2 rows)
```

So, overall, if you can point me the reason why we need these colocation checks, I'd be happy to provide input. But, my current thinking is that we should most likely undo all of these.

Reply:
As we've discussed, this change has been reverted.